src/models/base_model.py

```diff
@@ -0,0 +1,245 @@
+from abc import ABC, abstractmethod
+import torch
+import torch.nn as nn
+
+from dataclasses import dataclass
+from functools import partial
+from collections import OrderedDict, defaultdict
+from copy import deepcopy
+from typing import Dict, List, Tuple, Union, Generator, Optional, Any
+from enum import Enum
+
+
+class HookType(Enum):
+    FORWARD = 'forward'
+    FORWARD_PRE = 'forward_pre'
+
+
+@dataclass
+class FeatureInfo:
+    """Gather feature hooks params.
+    
+    Args:
+        module_name: Name of hooked module.
+        num_channels: Number of image channels.
+        stride: Downscale value for get image shape in current module.
+    """
+    module_name: str
+    num_channels: int
+    stride: int
+    hook_type: HookType = HookType.FORWARD
+
+
+@dataclass
+class Hook:
+    """Gather hook params.
+    
+    Args:
+        module_name: Hooked module name.
+        hook_type: Hook type (forward hook or forward pre hook).
+    """
+    module_name: str
+    hook_type: HookType
+
+
+class FeatureHooks:
+    """ Feature Hook Helper.
+
+    This module helps with the setup and extraction of hooks for extracting features from
+    internal nodes in a model by node name. This works quite well in eager Python but needs
+    redesign for torcscript.
+    """
+
+    def __init__(self, hooks: List[Hook], named_modules: Generator[Tuple[str, nn.Module], None, None]):
+        """Initialize feature hooks.
+        
+        Args:
+            hooks: Hooks to be registered.
+            named_modules: Result of self.named_modules() function call.
+        """
+        # setup feature hooks
+        modules = {k: v for k, v in named_modules}
+        for i, hook in enumerate(hooks):
+            hook_name = hook.module_name
+            module = modules[hook_name]
+            hook_fn = partial(self._collect_hook_features, hook_name)
+            # register hooks
+            if hook.hook_type == HookType.FORWARD_PRE:
+                module.register_forward_pre_hook(hook_fn)
+            else:
+                module.register_forward_hook(hook_fn)
+           
+        self._feature_outputs = defaultdict(OrderedDict)
```

Comment:
Define `self.__feature_hooks` and `self.__features_info` because every field of the class should be defined in `__init__`

Reply:
I don't get the point. While you are writing to the variable `self.*` you actually store the values in your class's field